libs/threading/tests/unit/stop_tokenrace.cpp

```diff
@@ -0,0 +1,238 @@
+    }
+};
+
+void test_callback_concurrent_unregister()
+{
+    // create stop_source and stop_token:
+    hpx::stop_source ssrc;
+    hpx::stop_token stok{ssrc.get_token()};
+
+    std::atomic<bool> cb1_called{false};
+    hpx::util::optional<hpx::stop_callback<std::function<void()>>> opt_cb;
+
+    auto cb1 = [&] {
+        opt_cb.reset();
+        cb1_called = true;
+    };
+
+    opt_cb.emplace(stok, std::ref(cb1));
+
+    // request stop
+    ssrc.request_stop();
+
+    HPX_TEST(ssrc.stop_possible());
+    HPX_TEST(ssrc.stop_requested());
+    HPX_TEST(stok.stop_possible());
+    HPX_TEST(stok.stop_requested());
+
+    HPX_TEST(cb1_called);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+void test_callback_concurrent_unregister_other_thread()
+{
+    // create stop_source and stop_token:
+    hpx::stop_source ssrc;
+    hpx::stop_token stok{ssrc.get_token()};
+
+    std::atomic<bool> cb1_called{false};
+    hpx::util::optional<hpx::stop_callback<std::function<void()>>> opt_cb;
+
+    auto cb1 = [&] {
+        opt_cb.reset();
+        cb1_called = true;
+    };
+
+    hpx::thread t{[&] { opt_cb.emplace(stok, std::ref(cb1)); }};
+
+    // request stop
+    ssrc.request_stop();
+
+    t.join();
+
+    HPX_TEST(ssrc.stop_possible());
+    HPX_TEST(ssrc.stop_requested());
+    HPX_TEST(stok.stop_possible());
+    HPX_TEST(stok.stop_requested());
+
+    HPX_TEST(cb1_called);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// void test_callback_throw()
+// {
+//     // create stop_source
+//     hpx::stop_source ssrc;
+//     HPX_TEST(ssrc.stop_possible());
+//     HPX_TEST(!ssrc.stop_requested());
+//
+//     // create stop_token from stop_source
+//     hpx::stop_token stok{ssrc.get_token()};
+//     HPX_TEST(ssrc.stop_possible());
+//     HPX_TEST(!ssrc.stop_requested());
+//     HPX_TEST(stok.stop_possible());
+//     HPX_TEST(!stok.stop_requested());
+//
+//     // register callback
+//     bool cb1_called{false};
+//     bool cb2_called{false};
+//
+//     auto cb = [&] {
+//         cb1_called = true;
+//         // throw
+//         throw "callback called";
+//     };
+//
+//     auto cb1 = hpx::make_stop_callback(stok, cb);
+//
+//     HPX_TEST(ssrc.stop_possible());
+//     HPX_TEST(!ssrc.stop_requested());
+//     HPX_TEST(stok.stop_possible());
+//     HPX_TEST(!stok.stop_requested());
+//     HPX_TEST(!cb1_called);
+//     HPX_TEST(!cb2_called);
+//
+//     // catch terminate() call:
+//     std::set_terminate([] { std::exit(0); });
+//
+//     // request stop
+//     ssrc.request_stop();
+//     HPX_TEST(false);
+// }
+
+///////////////////////////////////////////////////////////////////////////////
+int main()
+{
+    test_callback_register();
+    test_callback_unregister();
+
+    test_callback_concurrent_unregister();
+    test_callback_concurrent_unregister_other_thread();
+
+    // this test terminates execution
+    // test_callback_throw();
```

Comment:
Is this intentionally disabled?

Reply:
Yes, we have no way of testing whether `std::terminate` was actually called. Any ideas?

Reply:
Hmm, is it just that last `request_stop` (line 221) that would terminate? I don't know if this is a good idea, but does one have to call `exit` in the termination handler? If not that could be one option (not sure how messy the synchronization would be). Otherwise maybe separating out that test and printing something unique in the termination handler, and making CMake check for that output?

Reply:
If the callbacks throw an exception, `std::terminate` will be called (see [C++20](http://wg21.link/n4849), 32.3.5.1.4). 

Using CMake to detect some unique output would work.